### PR TITLE
Amp cluster management

### PR DESCRIPTION
**Pull request by freignat91**
Related to #512 

For the ones granted, here is a video to explain more in details the PR features
https://drive.google.com/a/axway.com/file/d/0B3HdJx5HHXGcUzR1ZVRSMGVURVk/view?usp=sharing

see also the discussion on the topic "Short demo on amp cluster features" on Mobilize

to test locally:
`make install`
`./shrink.sh local`
`ampadm pf pull --local -v`
`ampadm pf start --local -v`

`ampadm pf monitor` (on another console)
`ampadm node ls`
`ampadm node count`
`ampadm pf stop`

to test on Amazon amp cluster after having used the Nicolas's amp-bootstrap installation:
On the manager execute:
`ampadm cluster start --tag cluster`
(it use a predefined amp:cluster image for PR need, but after merge and image appelerator/amp:latest built and pulled no need of the option "--tag cluster")
That's all on manager.

on your laptop client machine execute:
`ampadm --server [managerAddress]:31315 node ls`  (to see the nodes status)
`ampadm --server [managerAddress]:31315 node purge --container --image` (if needed)
`ampadm --server [managerAddress]:31315 pf pull -v` (all amp images are pulled on all nodes)
`ampadm --server [managerAddress]:31315 pf start -v`
`ampadm --server [managerAddress]:31315 pf monitor` (on another console)

[managerAdresse] can be an aws load-balancer address

For now the previous amp pf commands stay available, waiting for decision about amp/amp-cluster cli integration or not.

-ampadm --server [managerAddress]:31315 pf pull -v

### CODE REVIEW

**Review by generalhenry — CHANGES_REQUESTED** [review]
Docker: 1.13.0-rc2 

freshly restored to factory defaults

`docker swarm init`
`make install`
`./shrink.sh local`
`amp-cluster pf start --local -v`

Result:

Service cluster-agent starting
Service cluster-server starting
Cluster services started
Connection to cluster server error: grpc: timed out when dialing

Reason:

my ~/.amp.yaml contains a normal serverAddress entry

The server address configuration needs to be independently set

File: cmd/amp-cluster/cli.go [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+		Run: func(cmd *cobra.Command, args []string) {
+			g.printfc(colSuccess, "amp-cluster version: %s, build: %s)\n", Version, Build)
+		},
+	}
+	RootCmd.AddCommand(versionCmd)
+
+	//Execute commad
+	cmd, _, err := RootCmd.Find(os.Args[1:])
+	if err != nil {
+		g.printfc(colError, "Arg error: %v\n", err)
```
Comment (generalhenry) [inline cmd/amp-cluster/cli.go]: The pattern of `g.printfc(colError, errorDescription, err); os.exit(1)` is common enough make a `g.fatalc(errorDescription, err)` method

File: cmd/amp-cluster/cli_node_count.go [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+
+func (g *clusterClient) containerCount(cmd *cobra.Command, args []string) error {
+	req := &servergrpc.GetNodesInfoRequest{}
+	follow := false
+	req.Node = cmd.Flag("node").Value.String()
+	if cmd.Flag("follow").Value.String() == "true" {
+		follow = true
+	}
+	if follow {
+		fmt.Println("\033[2J\033[0;0H")
```
Comment (generalhenry) [inline cmd/amp-cluster/cli_node_count.go]: These printlns are both arcane and used a number of places maybe a `g.followStart(follow bool)` and `g.followLn(follow bool)`

File: cmd/cluster-server/servergrpc/cluster-server.proto [hunk 3]
```diff
@@ -0,0 +1,128 @@
+syntax = "proto3";
+
+package servergrpc;
+
+import "google/api/annotations.proto";
+
+service ClusterServerService {
+        rpc GetClientStream (stream ClientMes) returns (stream ClientMes) {}
```
Comment (generalhenry) [inline cmd/cluster-server/servergrpc/cluster-server.proto]: grpc-gw annotations would be nice

**Comment by freignat91** [comment]
About server address configuration, i was waiting the decisions to merge or not the cluster-cli into amp-cli or cluster-server into amplifier to finalize the configuration (see the discussion topic "Short demo on amp cluster features" on Mobilize), but in the meantime i can make it separated yes.
Done: added `ClusterServerAddress` item in configuration

File: cmd/amp-cluster/cli.go (hunk 1 above)
Comment (freignat91) [reply cmd/amp-cluster/cli.go]: yes, why not. I prefer explicit Exit command, than hidden one, but to be aligned with Go style, it's ok.

File: cmd/amp-cluster/cli_node_count.go (hunk 2 above)
Comment (freignat91) [reply cmd/amp-cluster/cli_node_count.go]: done

File: cmd/amp-cluster/cli.go (hunk 1 above)
Comment (freignat91) [reply cmd/amp-cluster/cli.go]: done

File: cmd/cluster-server/servergrpc/cluster-server.proto (hunk 3 above)
Comment (freignat91) [reply cmd/cluster-server/servergrpc/cluster-server.proto]: yes sure, but first get agreement on the cli verbs/structure and inclusion or not in amp cli.

**Review by generalhenry — APPROVED** [review]

**Comment by subfuzion** [comment]
@freignat91 @ndegory It's taken a while for me to review, but excellent work. Thank you for this contribution. 👍 

I'm merging this into v0.5.0-rc1 for now for additional evaluation.

**Pull request merged**